authentication/authentication-common/src/main/java/io/mosip/authentication/common/service/integration/OTPManager.java

```diff
@@ -160,29 +200,104 @@ private String generateOTP(String uin) throws IdAuthUncheckedException {
 	 */
-	public boolean validateOtp(String pinValue, String otpKey) throws IdAuthenticationBusinessException {
-		String otpHash;
-		otpHash = IdAuthSecurityManager.digestAsPlainText(
-				(otpKey + EnvUtil.getKeySplitter() + pinValue).getBytes());
-		Optional<OtpTransaction> otpTxnOpt = otpRepo.findByOtpHashAndStatusCode(otpHash, IdAuthCommonConstants.ACTIVE_STATUS);
-		if (otpTxnOpt.isPresent()) {
-			OtpTransaction otpTxn = otpTxnOpt.get();
-			//OtpTransaction otpTxn = otpRepo.findByOtpHashAndStatusCode(otpHash, IdAuthCommonConstants.ACTIVE_STATUS);
-			otpTxn.setStatusCode(IdAuthCommonConstants.USED_STATUS);
-			otpRepo.save(otpTxn);
-			if (otpTxn.getExpiryDtimes().isAfter(DateUtils.getUTCCurrentDateTime())) {
-				return true;
-			} else {
+	public boolean validateOtp(String pinValue, String otpKey, String individualId) throws IdAuthenticationBusinessException {
+		String refIdHash = securityManager.hash(individualId);
+		Optional<OtpTransaction> otpEntityOpt = otpRepo.findFirstByRefIdOrderByGeneratedDtimesDesc(refIdHash);
+
+		requireKeyNotFound(otpEntityOpt);
+		
+		OtpTransaction otpEntity = otpEntityOpt.get();
+		
+		requireOtpNotFrozen(otpEntity);
+		
+		// Increment the validation attempt count.
+		int attemptCount = otpEntity.getValidationRetryCount() == null ? 1 : otpEntity.getValidationRetryCount() + 1;
+		
+		if(!otpEntity.getStatusCode().equals(IdAuthCommonConstants.ACTIVE_STATUS)) {
+			throw new IdAuthenticationBusinessException(IdAuthenticationErrorConstants.OTP_VAL_KEY_NOT_FOUND);
+		}
+		
+		String otpHash = getOtpHash(pinValue, otpKey);
+		if (otpEntity.getOtpHash().equals(otpHash)) {
+			otpEntity.setUpdDTimes(DateUtils.getUTCCurrentDateTime());
+			otpEntity.setStatusCode(IdAuthCommonConstants.USED_STATUS);
+			otpRepo.save(otpEntity);
+			if (!otpEntity.getExpiryDtimes().isAfter(DateUtils.getUTCCurrentDateTime())) {
 				logger.error(IdAuthCommonConstants.SESSION_ID, this.getClass().getSimpleName(),
 						IdAuthenticationErrorConstants.EXPIRED_OTP.getErrorCode(), OTP_EXPIRED);
 				throw new IdAuthenticationBusinessException(IdAuthenticationErrorConstants.EXPIRED_OTP);
 			}
+			return true;
 		} else {
+			//Set the incremented validation attempt count
+			otpEntity.setValidationRetryCount(attemptCount);
+			if (attemptCount >= numberOfValidationAttemptsAllowed) {
+				otpEntity.setStatusCode(IdAuthCommonConstants.FROZEN);
+			}
+			otpEntity.setUpdDTimes(DateUtils.getUTCCurrentDateTime());
+			otpRepo.save(otpEntity);
 			return false;
 		}
 	}
+
+	/**
+	 * Require otp not frozen.
+	 *
+	 * @param otpEntity the otp entity
+	 * @throws IdAuthenticationBusinessException the id authentication business exception
+	 */
+	private void requireOtpNotFrozen(OtpTransaction otpEntity) throws IdAuthenticationBusinessException {
+		if(otpEntity.getStatusCode().equals(IdAuthCommonConstants.FROZEN)) {
+			if(otpEntity.getUpdDTimes() == null || !isAfterFrozenDuration(otpEntity)) {

```

Comment:
not sure which scenario the updDTimes will be null.

Reply:
Updtimes will be null when the OTP is requested freshly, where only crDTimes will be set.